Add FullDeviceInfo trait and impl

dmarcuse opened the pull request:
This implements #104 for devices. A new trait, `ocl::FullDeviceInfo` (and matching implementation for `ocl::Device`) is added, with wrapper methods for every standard type of device info. For example, this code:
```rust
match device.info(DeviceInfo::DeviceType)? {  DeviceInfoResult(ty) => ..., _ => panic!(...) }
```
...can be rewritten as `device.device_type()?` when using `FullDeviceInfo`, greatly simplifying usage.

Right now I've only done manual testing of some of the methods, but they seem to work properly. Automatic tests would be nice, but I've been having trouble with running the tests on my system.

apemanzilla commented:
I know that the original intent was to replace `DeviceInfoResult` entirely, but for now, this is more straightforward, and retains backwards compatibility. If you still want to get rid of it, I think the best way to implement it would be to have a trait like the one below, and implementations for each type that can be returned from an OpenCL info query.

```rust
trait ParseOpenclInfo {
    fn parse_ocl_info(bytes: Vec<u8>) -> Self;
}
```

c0gent commented:
I think this is an interesting solution. I like it. Give me some time to consider all of the implications to the overall API. I'll probably want to add `Full...Info` traits for other types too if we merge this.

I don't necessarily *want* to get rid of `DeviceInfoResult`, it's just sort of a cumbersome API. I don't think it's terribly practical to get rid of it though, considering all of the extra code bloat it would require.

apemanzilla commented:
I'm not sure if we need a `Full___Info` for _every_ OpenCL type, many of them (platforms, events, contexts, kernels, etc) only have a few info parameters, and for some (e.g. platforms) these are already implemented as methods on the struct itself.

c0gent commented:
Good point.

c0gent commented:
I'm still not sure that this is worth adding. I'd like more feedback from others before I merge.

apemanzilla commented:
Any reason in particular? It greatly simplifies reading device info that doesn't already have a dedicated method.

c0gent commented:
Mainly that additional code (especially macros) has an impact on compile times. It may seem small but it adds up fast.

I think this is a good idea and I would like to merge this but I'll only do so if others find it useful.

dmarcuse commented:
If this is still open for discussion, I'd like to point out that the compilation time is (for most use cases) a one-time cost due to incremental compilation. Rebuilds will occur when the package is updated of course, but those are relatively infrequent.

c0gent commented:
Good point. 

Yes actually, perhaps `ocl-extras` might be a good place. If you move it there I'd gladly merge that and then we can figure out if moving it into `ocl` at some point in the future makes sense.

Go ahead and beef up the documentation also if you would, explaining the use case in more detail and providing an example or two.

dmarcuse commented:
I've moved it to `ocl-extras` and added better documentation. I messed around with the commit history a bit to avoid ending up with a giant mess, so you'll probably need to force pull or re-clone the branch if you have a local copy.

c0gent commented:
Ok great. Thanks :)

dmarcuse commented:
Anything else I should do for this to be merged?

c0gent commented:
I just didn't have time to test it out last night. Looks good!

dmarcuse commented:
Thank you!

c0gent commented:
[0.1.1](https://crates.io/crates/ocl-extras) published :)

dmarcuse commented:
Perfect, thanks again!
